### PR TITLE
Modularization guidelines

### PR DESCRIPTION
**Pull request by BongaTheProto**
## About The Pull Request
Adds guidelines on how to proceed with modularization in our codebase, based loosely on skyrat's modularization guidelines

## Why It's Good For The Game
Skyrat has their own so we may as well create our guidelines to avoid confusion

## Proof Of Testing
I mean, you can read it, right?

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -23,13 +23,11 @@ Space Station 13 is a paranoia-laden round-based roleplaying game set against th
 ## Contribution Rules and Guidelines
 
 **1. Do Not Be A Dick**
-- The S.P.L.U.R.T. main repository is run by and contributed by volunteers. You are not entitled to our time and energy. We reserve the right to permanently remove anyone who does not show both our contributor's and maintainer's common decency.
+- The S.P.L.U.R.T. main repository is run by and contributed by volunteers. Please be considerate with the people who help maintaining our station. We reserve the right to permanently remove anyone who does not show both our contributor's and maintainer's common decency.
```
Comment (liz-lavenza) [inline README.md:26]:
```suggestion
- The S.P.L.U.R.T. main repository is run by and contributed by volunteers. Please be considerate with the people who help maintain our codebase. We reserve the right to permanently remove anyone who does not show both our contributors and maintainers common decency.
```

File: README.md [hunk 2]
```diff
@@ -42,7 +40,6 @@ Space Station 13 is a paranoia-laden round-based roleplaying game set against th
 - You are entitled to credit yourself with comments and you are entitled to waive the attribution requirement choosing not to be identified as the creator if you wish.
 - If you do not like how your assets were modified or used, it is required that the other person [remove the attribution information upon request](https://wiki.creativecommons.org/wiki/License_Versions#Licensors_may_request_removal_of_attribution).
 - Modifications or adaptions must disclose the source, the author, and [any changes you've made](https://wiki.creativecommons.org/wiki/License_Versions#Modifications_and_adaptations_must_be_indicated).
-- Goonstation code is incompatible with this codebase and will not be accepted.
```
Comment (liz-lavenza) [inline README.md:45]: Why was this removed? Goonstation code cannot be used, it's licensed under a Creative Commons Share-Alike license, which is incompatible with the GPL and AGPL families of licenses.

File: modular_zzplurt/readme.md [hunk 3]
```diff
@@ -0,0 +1,165 @@
+# The Modularization Handbook - S.P.L.U.R.T. Style, v1.0
+
+## Introduction
+
+Welcome to the S.P.L.U.R.T. station's modularization handbook! Our goal is to make contributing to our codebase as easy and comfy as possible for coders while upkeeping our code standards. We understand that maintaining a codebase that's a fork of another can be challenging, but with the right practices, we can keep our code clean, organized, and easy to manage. This handbook outlines our modularization protocols and coding standards to help you get started.
```
Comment (liz-lavenza) [inline modular_zzplurt/readme.md:5]:
```suggestion
Welcome to the S.P.L.U.R.T. codebase's modularization handbook! Our goal is to make contributing to our codebase as easy and comfy as possible for coders while upkeeping our code standards. We understand that maintaining a codebase that's a fork of another can be challenging, but with the right practices, we can keep our code clean, organized, and easy to manage. This handbook outlines our modularization protocols and coding standards to help you get started.
```
I'm a little nitpicky about calling it the SPLURT codebase rather than the SPLURT station (the station's in-game and created by mappers, after all), so feel free to ignore if it's not to your tastes.

File: modular_zzplurt/readme.md [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+
+Our modularization protocols are founded on three pillars:
+
+1. **Modularize anything that's reasonable to modularize:** This includes anything that can be modularized without decreasing its quality or the quality of existing code, and won't logically cause more issues by modularizing than not doing it.
+2. **Use commenting conventions for non-modular edits:** When editing non-modular code, make sure to adequately use the commenting conventions to ensure it is known that it's a non-modular edit.
```
Comment (liz-lavenza) [inline modular_zzplurt/readme.md:44]:
```suggestion
2. **Use commenting conventions for non-modular edits:** When editing non-modular code, make sure to adequately use the [commenting conventions](#commenting-conventions) to ensure it is known that it's a non-modular edit.
```
This should add a link to the relevant section in this document. I think.

File: modular_zzplurt/readme.md [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+3. **It's ok to mess up:** These rules are intended less as a set of guidelines for contributors to strictly follow, but rather coding standards that we all collectively help to attain, for examble through maintainers helping any contributors fulfill them in their pull requests.
```
Comment (liz-lavenza) [inline modular_zzplurt/readme.md:45]:
```suggestion
3. **It's ok to mess up:** These guidelines are intended less as a set of ironclad rules for contributors to strictly follow, but rather coding standards that we all collectively help to attain, for example through maintainers helping any contributors fulfill them in their pull requests.
```
I would call these guidelines rather than rules, personally, but that's again a matter of personal taste in writing. Also, typo; `examble` should be `example`.

File: modular_zzplurt/readme.md [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+
+In this example, the shoot_live_shot proc is overridden to add a new behavior (spawning sparks) after calling the original proc. This ensures that the new behavior is added without modifying the core code.
+
+#### When to Use Modular Overrides
+
+Modular overrides should be used whenever possible to keep the core codebase clean and maintainable. However, it's important to use them judiciously and ensure that they do not introduce performance issues or make the code harder to understand and maintain. Specifically:
+
+- **Use modular overrides when they won't cause more issues than not modularizing them:** If a modular override would introduce significant complexity, performance issues, or maintenance challenges, it may be better to make a non-modular edit with proper commenting.
+- **Avoid unnecessary overrides:** Do not override procs or variables unless it is necessary for your feature or fix. Unnecessary overrides can lead to maintenance headaches and potential conflicts with upstream changes.
+- **Document your overrides:** Clearly documenting any modular overrides in your code and helps maintainers and other contributors understand the changes and their purpose.
```
Comment (liz-lavenza) [inline modular_zzplurt/readme.md:86]:
```suggestion
- **Document your overrides:** Clearly documenting any modular overrides in your code helps maintainers and other contributors understand your changes and their purpose.
```

File: modular_zzplurt/readme.md [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+```
+
+In this example, the shoot_live_shot proc is overridden to add a new behavior (spawning sparks) after calling the original proc. This ensures that the new behavior is added without modifying the core code.
+
+#### When to Use Modular Overrides
+
+Modular overrides should be used whenever possible to keep the core codebase clean and maintainable. However, it's important to use them judiciously and ensure that they do not introduce performance issues or make the code harder to understand and maintain. Specifically:
+
+- **Use modular overrides when they won't cause more issues than not modularizing them:** If a modular override would introduce significant complexity, performance issues, or maintenance challenges, it may be better to make a non-modular edit with proper commenting.
+- **Avoid unnecessary overrides:** Do not override procs or variables unless it is necessary for your feature or fix. Unnecessary overrides can lead to maintenance headaches and potential conflicts with upstream changes.
```
Comment (liz-lavenza) [inline modular_zzplurt/readme.md:85]: Is `Do not override procs or variables unless it is necessary for your feature or fix.` really worth saying? I feel like everyone adding something thinks it's necessary; it'd be better to say something like
```suggestion
- **Avoid unnecessary overrides:** Do not override procs or variables unless it is the only reasonable way to implement your feature or fix. Unnecessary overrides can lead to maintenance headaches and potential conflicts with upstream changes.
```
That ties in nicely with the first pillar in `## The Modularization Protocol` above, in that modular edits and overrides should only be used when they're the most reasonable option. We could even add a link to that section here (or have an entirely new section explaining what it means for a change to be reasonable, if necessary).

File: modular_zzplurt/readme.md [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+
+```byond
+/obj/item/gun/shoot_live_shot(mob/living/user, pointblank = 0, atom/pbtarget = null, message = 1)
+    . = ..() // Call the original proc and set its value to the default return value
+    if(muzzle_flash)
+        spawn_sparks(src) // Add new behavior
+```
+
+In this example, the shoot_live_shot proc is overridden to add a new behavior (spawning sparks) after calling the original proc. This ensures that the new behavior is added without modifying the core code.
+
```
Comment (liz-lavenza) [inline modular_zzplurt/readme.md:79]: We should probably add something here (or somewhere else in the document, with a link) about file include order; proc overrides run in the order of last-defined to first-defined, and BYOND orders files in alphabetical order (with `_` being before all letters, and `~` being after). This would also potentially be a good spot to explain why the folder is `modular_zzplurt` rather than `modular_splurt`; `modular_zz` ensures that it comes after all the other `modular_` folders.

File: modular_zzplurt/readme.md [hunk 9]
```diff
@@ -0,0 +1,165 @@
...
+- **Document your overrides:** Clearly documenting any modular overrides in your code and helps maintainers and other contributors understand the changes and their purpose.
+By following these guidelines, we can ensure that our codebase remains clean, maintainable, and easy to navigate, while still allowing for the flexibility to add new features and make necessary changes.
+
+## Folder Structure
+
+Instead of using different modules for every different bit of the game, we prefer to use a structure that resembles the structure of the repo itself inside our modular folder. This means we have one coding folder for all the code, one icons folder, one sounds folder, etc.
+
+- **Code:** Any .dm files should go in `modular_zzplurt/code/`.
+- **Icons:** Any .dmi files should go in `modular_zzplurt/icons/`.
+- **Sounds:** Any sound files should go in `modular_zzplurt/sounds/`.
```
Comment (liz-lavenza) [inline modular_zzplurt/readme.md:95]: We should probably make a note about what reasonable sizes and formats for sound files are—we don't want to bloat the rsc with massive uncompressed `.wav` files. Everything added should be a `.ogg` file with as much compression as possible without hurting audio quality too severely, unless you have an *extremely* good reason otherwise.

**Review by liz-lavenza — CHANGES_REQUESTED** [review]
Some thoughts and feedback. I included proofreading, formatting, and phrasing changes in here, so feel free to ignore any of those if they don't strike your fancy, but I did have some ideas for how to make the guidelines clearer and for a few more topics to cover in passing.

File: modular_zzplurt/readme.md (hunk 7 above)
Comment (BongaTheProto) [reply modular_zzplurt/readme.md:85]: Yeah, the idea that the modularization protocols should convey to coders in that regard is "Always try first modularizing your changes. Only edit non modular code if it's not possible to modularize or if modularizing would worsen the code compared to not modularizing."

File: modular_zzplurt/readme.md (hunk 8 above)
Comment (BongaTheProto) [reply modular_zzplurt/readme.md:79]: Oh yeah that makes sense, it'd be a good idea to explain the order of including and compilation in byond. I think it'd be better to put it under this section

File: modular_zzplurt/readme.md (hunk 9 above)
Comment (BongaTheProto) [reply modular_zzplurt/readme.md:95]: Yeah, that could go in this section or on this section explaining binary files
